### PR TITLE
feat(frontend): convert order-by & limit to TopN in optimizer

### LINKED ISSUE #1459: frontend: support stream sort / TopN operator

### PR DESCRIPTION
**Pull request by fuyufjh**
## What's changed and what's your intention?

1. Converts queries with `order by` & `limit` to a TopN node in the optimizer.
2. For batch queries, TopN will be converted to `BatchTopN` **without** additional `Sort`
3. For streaming queries, TopN will be converted to `StreamTopN` while the `Materialize` still need to be aware of the ordering.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

close #1459

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/order_by.yaml [hunk 1]
```diff
@@ -43,3 +47,27 @@
...
   binder_error: 'Item not found: output column "v"'
+- sql: |
+    create table t (v1 bigint, v2 double precision);
+    select * from t order by v1 desc limit 5;
+  batch_plan: |
+    BatchTopN { order: [$0 DESC] limit: 5, offset: 0 }
+      BatchExchange { order: [], dist: Single }
+        BatchScan { table: t, columns: [v1, v2] }
+  stream_plan: |
+    StreamMaterialize { columns: [v1, v2, _row_id#0(hidden)], pk_columns: [v1] }
```
Comment (fuyufjh) [inline src/frontend/test_runner/tests/testdata/order_by.yaml:58]: The PK derivation of TopN needs to be fixed in later PRs.

File: src/frontend/test_runner/tests/testdata/order_by.yaml [hunk 2]
```diff
@@ -43,3 +47,27 @@
     create table t (v1 bigint, v2 double precision);
     select * from t order by v;
   binder_error: 'Item not found: output column "v"'
+- sql: |
+    create table t (v1 bigint, v2 double precision);
+    select * from t order by v1 desc limit 5;
+  batch_plan: |
+    BatchTopN { order: [$0 DESC] limit: 5, offset: 0 }
```
Comment (skyzh) [inline src/frontend/test_runner/tests/testdata/order_by.yaml:54]: missing a `,` after order field?

File: src/frontend/test_runner/tests/testdata/order_by.yaml (hunk 1 above)
Comment (skyzh) [inline src/frontend/test_runner/tests/testdata/order_by.yaml:58]: Why we only have `v1` in pk_columns? If there are two elements of the same value from topN, we should be able to distinguish them.

**Review by skyzh — COMMENTED** [review]
Rest LGTM, good work! Also we may try enabling more TPC-H stream_plan in planner test.

File: src/frontend/test_runner/tests/testdata/order_by.yaml (hunk 1 above)
Comment (fuyufjh) [reply src/frontend/test_runner/tests/testdata/order_by.yaml:58]: Yes, as I had commented:

> The PK derivation of TopN needs to be fixed in later PRs.

**Review by skyzh — APPROVED** [review]

**Review by lmatz — APPROVED** [review]

**Comment by fuyufjh** [comment]
Oops.. the PK derivation problem is caused by a stupid bug made by me... Fixed now!

**Pull request merged**